impeller/tessellator/tessellator.cc

```diff
@@ -237,9 +237,10 @@ EllipticalVertexGenerator::EllipticalVertexGenerator(
 EllipticalVertexGenerator Tessellator::FilledCircle(
     const Matrix& view_transform,
     const Point& center,
+    Scalar raw_radius,
     Scalar radius) {
-  auto divisions =
-      ComputeQuadrantDivisions(view_transform.GetMaxBasisLength() * radius);
+  size_t divisions = ComputeQuadrantDivisions(
+      view_transform.GetMaxBasisLengthXY() * raw_radius);
```

Comment:
If the scaled radius is very small then we choose very few points to represent it, but the actual dimensions of what is drawn might be large enough to see the loss of precision.

On the other hand, these are all 1-pixel in size at this point? Is it worth it either way? Since we've fixed the issue with the determinant, how much does any of this matter any more?

Reply:
This is a pretty critical part of the bug fix, the scenario:

GetMaxBasisXY = 1e20
raw_radius = 1e-20

If we use the rounded up radius, we compute divisions as if we had a radius of 1e20, if we use the raw radius we get the appropriate divisions of ~1.

Reply:
How so? The calling code would have done:

min_size = 1/MaxBasis = 1/1e20 = 1e-20
radius = max(radius, min_size) == max(1e-20, 1e-20) == 1e-20

Calling the tessellator it would have then computed 1e-20 * MaxBasis == 1e-20 * 1e20 == 1.0 and computed divisions for that.

Reply:
The "rounded up radius" should never be greater than 1.0 when multiplied by MaxBasisXY.

Basically, the calling code is saying "When we ask for the tesselation it will base it on (radius * MaxBasis). Let's make sure that product is >= 1.0. To do that the radius can never be smaller than (1/MaxBasis)."

Another way of doing this is:

scaled_radius = radius * MaxBasis
if (scaled_radius < 1.0) scaled_radius = 1.0
scaled_radius /= MaxBasis.

Then hand that value into the tesselator and it will compute "passed_radius * MaxBasis" and that value can never be less than 1.0.

The code in the function is basically the same thing: max(radius, 1/MaxBasis) computes the same value.

Reply:
The actual problems were that it was using the determinant which is "similar", but not the same value (it is the area of the unit square, not the largest scale factor), and then the code in the function was using the 3D version of MaxBasis. Does it not work if you replace both with MaxBasisXY?

Reply:
You're right! agh, total facepalm on my end.